File: Java/interview_33.java
class Solution {
    private boolean isValid(int[] postorder, int left, int right) {
        if (left >= right) {
            return true;
        }
        int root = postorder[right];
        int idx;
        for (idx = left; idx < right; ++idx) {
            if (postorder[idx] > root) {
                break;
            }
        }
        for (int i = idx; i < right; ++i) {
            if (postorder[i] < root) {
                return false;
            }
        }
        return isValid(postorder, left, idx - 1) && isValid(postorder, idx, right - 1);
    }
    public boolean verifyPostorder(int[] postorder) {
        return isValid(postorder, 0, postorder.length - 1);
    }
}

File: python/interview_33.py
class Solution:
    def verifyPostorder(self, postorder: List[int]) -> bool:
        def isValid(left, right):
            if left >= right:
                return True
            i = left
            while i < right:
                if postorder[i] > postorder[right]:
                    break
                i += 1
            j = i
            while j < right:
                if postorder[j] < postorder[right]:
                    return False
                j += 1
            return isValid(left, i - 1) and isValid(i, right - 1)
        return isValid(0, len(postorder) - 1)
